ec/src/models/bls12/mod.rs

```diff
@@ -145,35 +149,46 @@ impl<P: Bls12Parameters> PairingEngine for Bls12<P> {
             // r = f^((p^6 - 1)(p^2 + 1))
             r *= &f2;
 
-            // Hard part of the final exponentation is below:
-            // From https://eprint.iacr.org/2016/130.pdf, Table 1
+            // Hard part of the final exponentation:
+            // t[0].CyclotomicSquare(&result)
             let mut y0 = r.cyclotomic_square();
-            y0.conjugate();
-
-            let mut y5 = Self::exp_by_x(r);
-
-            let mut y1 = y5.cyclotomic_square();
-            let mut y3 = y0 * &y5;
-            y0 = Self::exp_by_x(y3);
-            let y2 = Self::exp_by_x(y0);
-            let mut y4 = Self::exp_by_x(y2);
-            y4 *= &y1;
-            y1 = Self::exp_by_x(y4);
-            y3.conjugate();
-            y1 *= &y3;
-            y1 *= &r;
-            y3 = r;
-            y3.conjugate();
-            y0 *= &r;
-            y0.frobenius_map(3);
-            y4 *= &y3;
-            y4.frobenius_map(1);
-            y5 *= &y2;
-            y5.frobenius_map(2);
-            y5 *= &y0;
-            y5 *= &y4;
-            y5 *= &y1;
-            y5
+            // t[1].Expt(&result)
+            let mut y1 = Self::exp_by_x(r);
+            // t[2].InverseUnitary(&result)
+            let mut y2 = r;
+            y2.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
+            // t[2].Expt(&t[1])
+            y2 = Self::exp_by_x(y1);
+            // t[1].InverseUnitary(&t[1])
+            y1.conjugate();
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= y2;
+            // t[2].Expt(&t[1])
+            y2 = Self::exp_by_x(y1);
+            // t[1].Frobenius(&t[1])
+            y1.frobenius_map(1);
+            // t[1].Mul(&t[1], &t[2])
+            y1 *= &y2;
+            // result.Mul(&result, &t[0])
+            r *= y0;
+            // t[0].Expt(&t[1])
+            y0 = Self::exp_by_x(y1);
+            // t[2].Expt(&t[0])
+            y2 = Self::exp_by_x(y0);
+            // t[0].FrobeniusSquare(&t[1])
+            y0 = y1;
```

Comment:
Ditto for the copy here

Reply:
Hm how do I avoid the copy of `y1` here?

Reply:
This is how I did it

```Nim
  # (x²+p²−1)
  v2.pow_x(v0, invert = false)
  v1.pow_x(v2, invert = false) # v1 = f^((x-1)².(x+p).x²)
  v2.frobenius_map(v0, 2)      # v2 = f^((x-1)².(x+p).p²)
  v0.cyclotomic_inv()          # v0 = f^((x-1)².(x+p).-1)
  v0 *= v1                     # v0 = f^((x-1)².(x+p).(x²-1))
  v0 *= v2                     # v0 = f^((x-1)².(x+p).(x²+p²-1))

  # (x−1)².(x+p).(x²+p²−1) + 3
  f *= v0
```